### PR TITLE
sap.sap_operations collection content

### PR DESCRIPTION
**Pull request by kksat**

### CODE REVIEW

File: plugins/modules/hana_backup.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+  Create backup of SAP HANA database to files or using BACKINT interface
+  Full Backup: complete SAP HANA database backup.
+  Differential Backup: the data changed since the last full data backup.
+  Incremental Backup: the data changed since the last full data backup or the last delta backup.
+  https://help.sap.com/docs/SAP_HANA_PLATFORM/6b94445c94ae495c83a19646e7c3fd56/15b4aa82ae7544f78f809e35add006ce.html
+
+version_added: 1.1.0
+
+seealso:
+  - module: sap.sap_operations.hana_backup
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:31]:
```suggestion
  - module: community.sap_operations.hana_backup
```

File: plugins/modules/hana_backup.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+short_description: Create SAP HANA database file backup
+
+description: |
+  Create backup of SAP HANA database to files or using BACKINT interface
+  Full Backup: complete SAP HANA database backup.
+  Differential Backup: the data changed since the last full data backup.
+  Incremental Backup: the data changed since the last full data backup or the last delta backup.
+  https://help.sap.com/docs/SAP_HANA_PLATFORM/6b94445c94ae495c83a19646e7c3fd56/15b4aa82ae7544f78f809e35add006ce.html
+
+version_added: 1.1.0
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:28]:
```suggestion
version_added: 1.0.0
```

File: plugins/modules/hana_backup.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright 2022 Red Hat, Project Atmosphere
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:5]: Must be apache with a exception to prevent lint fails

File: plugins/modules/hana_backup.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    database_user_password: CHANGEME
+    database_name: RHE
+    prefix: MONDAY
+    destination: /backups
+    comment: Created with ansible
+    binary_path: "/my/custom/path"
+    wait: false
+"""
+
+RETURN = r""" # """
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:139]: should have a reutrn example like:

```
RETURN = r'''
query_result:
    description: List containing results of all queries executed (one sublist for every query).
    returned: on success
    type: list
    elements: list
```

File: plugins/modules/hana_backup.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+      - This parameter is used only in case C(type) is I(BACKINT).
+    type: str
+"""
+
+EXAMPLES = r"""
+- name: Fetch the binary path of the hdbsql
+  sap.sap_operations.parameter_info:
+    instance_number: "00"
+    name: DIR_INSTANCE
+  register: dir_instance
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:113]: not part of the module it self

File: plugins/modules/hana_backup.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
+#!/usr/bin/python
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:1]: does it really make sense to write a complete module for a few sql statements ? maybe it's less maintaining effort use the hdbsql module from sap_libs and write a role

File: plugins/modules/hana_backup.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+EXAMPLES = r"""
+- name: Fetch the binary path of the hdbsql
+  sap.sap_operations.parameter_info:
+    instance_number: "00"
+    name: DIR_INSTANCE
+  register: dir_instance
+
+- name: Create a full HANA backup for SYSTEMDB
+  sap.sap_operations.hana_backup:
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:116]:
```suggestion
  community.sap_operations.hana_backup:
```

File: plugins/modules/hana_backup.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    database_user_password: CHANGEME
+    prefix: MONDAY
+    database_name: SYSTEMDB
+    binary_path: "{{ dir_instance.parameter_value | first }}"
+
+- name: Create a full HANA backup for RHE tenant
+  sap.sap_operations.hana_backup:
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:123]:
```suggestion
  community.sap_operations.hana_backup:
```

File: plugins/modules/hana_backup.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    database_user_password: CHANGEME
+    prefix: MONDAY
+    binary_path: "{{ dir_instance.parameter_value | first }}"
+
+- name: Create a full HANA backup for RHE tenant - all options used
+  sap.sap_operations.hana_backup:
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:129]:
```suggestion
  community.sap_operations.hana_backup:
```

File: plugins/modules/hana_backup.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+    module_args = dict(
+        host=dict(type="str"),
+        instance_number=dict(type="str"),
+        database_user_password=dict(type="str", no_log=True),
+        database_user=dict(type="str"),
+        hdbsqluserstore_key=dict(type="str", no_log=False),
+        database_name=dict(type="str", default="SYSTEMDB"),
+        type=dict(
+            type="str", choices=["INCREMENTAL", "DIFFERENTIAL", "FULL"], default="FULL"
+        ),
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:233]: should be outsourced to a own var outside the module_args to be clearer

Comment (rainerleber) [reply plugins/modules/hana_backup.py:233]: see: https://github.com/sap-linuxlab/community.sap_libs/blob/c4d27366c9b3b7f71f52409c25433f4f59c4843b/plugins/modules/sap_task_list_execute.py#L251 for example

File: plugins/modules/hana_backup.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+        comment=dict(type="str", default=""),
+        wait=dict(type="bool", default=True),
+        binary_path=dict(type="str", default=""),
+        backend=dict(type="str", choices=["FILE", "BACKINT"], default="FILE"),
+        tooloption=dict(type="str"),
+    )
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        supports_check_mode=True,
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:245]: it seams the module does not support a check mode

File: plugins/modules/hana_backup.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+    if rc != 0:
+        module.fail_json(cmd=args, rc=rc, stdout=stdout, msg=err)
+
+    return True, stdout, err, args
+
+
+def main():
+    module_args = dict(
+        host=dict(type="str"),
+        instance_number=dict(type="str"),
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:226]: required=true ?

File: plugins/modules/hana_backup.py [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+        database_user_password=dict(type="str", no_log=True),
+        database_user=dict(type="str"),
+        hdbsqluserstore_key=dict(type="str", no_log=False),
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:229]:
```suggestion
        hdbsqluserstore_key=dict(type="str"),
```
it's the default

File: plugins/modules/hana_backup.py [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+  Differential Backup: the data changed since the last full data backup.
+  Incremental Backup: the data changed since the last full data backup or the last delta backup.
+  https://help.sap.com/docs/SAP_HANA_PLATFORM/6b94445c94ae495c83a19646e7c3fd56/15b4aa82ae7544f78f809e35add006ce.html
+
+version_added: 1.1.0
+
+seealso:
+  - module: sap.sap_operations.hana_backup
+
+options:
```
Comment (rainerleber) [inline plugins/modules/hana_backup.py:33]: each parameter should contain required except there is a default

File: plugins/modules/hdbuserstore.py [hunk 15]
```diff
@@ -0,0 +1,273 @@
+#!/usr/bin/python
```
Comment (rainerleber) [inline plugins/modules/hdbuserstore.py:1]: apache as above

File: plugins/modules/hdbuserstore.py [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+# Do not use interactive flag when becoming <hanasid>adm user. But do not want to provide value for variable I(binary_path).
+#
+# In that case value for I(binary_path) can be extracted from HANA parameter DIR_EXECUTABLE that one can get with I(parameter_info) module:
+- name: Get DIR_EXECUTABLE
+  sap.sap_operations.parameter_info:
+    instance_number: "00"
+    name: DIR_EXECUTABLE
+  become: true
+  become_user: <hanasid>adm
+  register: __DIR_EXECUTABLE
```
Comment (rainerleber) [inline plugins/modules/hdbuserstore.py:156]: not part of the module

File: plugins/modules/hdbuserstore.py [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+# ansible modules fail. This is due to the fact that <hanasid>adm user sets environment variables PYTHONHOME and PYTHONPATH (to use HANA python,
+# not platform python) that confuse ansible.
+#
+# In that case hdbuserstore command will not be in PATH environment variable for <hanasid>adm user and I(binary_path) has to be provided.
+#
+# There are several workaround around this unplesant situation. One is recommended.
+#
+# Workaround 1 (recommended)
+#
+# Run hdbsuserstore module with <hanasid>adm user with '-i' (interactive) flag like so
```
Comment (rainerleber) [inline plugins/modules/hdbuserstore.py:97]: comments should not be part of the module

**Review by rainerleber — CHANGES_REQUESTED** [review]
first module checked see comments

File: plugins/modules/hdbuserstore.py [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+    support: full
+  diff_mode:
+    support: none
+  platform:
+    platforms: posix
+"""
+
+EXAMPLES = r"""
+- name: Set the key mykey
+  sap.sap_operations.hdbuserstore:
```
Comment (rainerleber) [inline plugins/modules/hdbuserstore.py:79]:
```suggestion
  community.sap_operations.hdbuserstore:
```

File: plugins/modules/hdbuserstore.py [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+  tasks:
+    - name: Environment for SAP HANA
+      set_fact:
+        ansible_python_interpreter: "/usr/libexec/platform-python -E"
+
+# Workaround 2
+#
+# Do not use interactive flag when becoming <hanasid>adm user.
+- name: Set the key mykey
+  sap.sap_operations.hdbuserstore:
```
Comment (rainerleber) [inline plugins/modules/hdbuserstore.py:134]:
```suggestion
  community.sap_operations.hdbuserstore:
```

File: plugins/modules/hdbuserstore.py (hunk 16 above)
Comment (rainerleber) [reply plugins/modules/hdbuserstore.py:156]: i know it's an example to get the binary but it's against the good practice instead you should setup examples

File: plugins/modules/hdbuserstore.py [hunk 20]
```diff
@@ -0,0 +1,273 @@
...
+- name: Get DIR_EXECUTABLE
+  sap.sap_operations.parameter_info:
+    instance_number: "00"
+    name: DIR_EXECUTABLE
+  become: true
+  become_user: <hanasid>adm
+  register: __DIR_EXECUTABLE
+
+- name: Set the key mykey
+  sap.sap_operations.hdbuserstore:
```
Comment (rainerleber) [inline plugins/modules/hdbuserstore.py:159]:
```suggestion
  community.sap_operations.hdbuserstore:
```

File: plugins/modules/hdbuserstore.py [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+    type: str
+  username:
+    description: |
+      Username for the hdb store
+      Required only if you set new key, C(state=present)
+    type: str
+  password:
+    description: |
+      Password for the hdb store username.
+      Required only if you set new key, state=present
```
Comment (rainerleber) [inline plugins/modules/hdbuserstore.py:57]:
```suggestion
      Required only if you set new key, C(state=present)
```

File: plugins/modules/hdbuserstore.py [hunk 22]
```diff
@@ -0,0 +1,273 @@
...
+    key = module.params.get("key")
+
+    rc, stdout, _err = module.run_command(args=[binary, "List", key])
+    if not module.check_mode and rc == 0:
+        rc, _out, _err = module.run_command(args=[binary, "Delete", key])
+
+    if rc not in [0, 100]:
+        module.fail_json(msg="Failed to execute delete: {0}".format(stdout))
+
+    return rc == 0, stdout, {"key": key}
```
Comment (rainerleber) [inline plugins/modules/hdbuserstore.py:231]:
```suggestion
    return True, stdout, {"key": key}
```

Comment (rainerleber) [reply plugins/modules/hdbuserstore.py:231]: otherwise the rc is used as result['changed']  see line 265

File: plugins/modules/hdbuserstore.py [hunk 23]
```diff
@@ -0,0 +1,273 @@
...
+        ],
+    )
+
+    state = module.params.get("state")
+    if state == "present":
+        result["changed"], result["stdout"], return_key = ensure_created(module)
+    else:
+        result["changed"], result["stdout"], return_key = ensure_absent(module)
+
+    result.update(return_key)
```
Comment (rainerleber) [inline plugins/modules/hdbuserstore.py:267]: what's the value of this command ? in my opinion line 262 or 265 should update the return_key

**Review by rainerleber — COMMENTED** [review]
hdbuserstore.py review

**Review by rhmk — COMMENTED** [review]
I basically agree to Rainers review from what I see. My biggest concern is the license incompatibility. Technically I suggest rejecting this particular PR and change the code with the renaming script/playbook to change from sap.sap_operations to community.sap_operations that Bernd uses to rename community to redhat. Then create a new PR that addresses the required changes from Rainer.

**Review by rhmk — COMMENTED** [review]
I basically agree to Rainers review from what I see. My biggest concern is the license incompatibility. Technically I suggest rejecting this particular PR and change the code with the renaming script/playbook to change from sap.sap_operations to community.sap_operations that Bernd uses to rename community to redhat. Then create a new PR that addresses the required changes from Rainer.

**Comment by sean-freeman** [comment]
Closing this PR for the following reasons:

1. PR is a copy/paste without any integration
    - to existing Ansible Collection sap_libs for low-level Ansible Modules
    - to existing Ansible Roles within this Ansible Collection
    - it is not acceptable to "dump" code and other community members are expected to test/integrate
2. PR has no description of contents
3. PR has no waiver of GPL License from the rogue duplicate named sap_operations Ansible Collection. All code within SLL-OSI is Apache to facilitate business re-use, with a code-to-product strategy using standardised Community First approach; e.g. upstream > midstream (e.g. fork repo in 'redhat-sap' for productisation amendments) > downstream
4. PR review comments provided on same day as PR, were never addressed
5. PR has been waiting over a year for the above activities and there is clear lack of community spirit / unwillingness to collaborate

**Pull request closed without merging**